File: lib/ts/src/serialize.test.ts
import {test} from 'vitest';
import {ChildIPC} from './child.js';

// Access serialize/deserialize through a ChildIPC instance's public methods
// We create a minimal wrapper to test them
class TestableIPC {
    private ipc: any;

    constructor() {
        // Access the prototype methods directly
        this.ipc = Object.create(ChildIPC.prototype);
    }

    serialize(arg: any): any {
        return this.ipc.serialize(arg);
    }

    deserialize(arg: any): any {
        return this.ipc.deserialize(arg);
    }
}

test('serialize primitives', ({expect}) => {
    const t = new TestableIPC();
    expect(t.serialize(42)).toBe(42);
    expect(t.serialize('hello')).toBe('hello');
    expect(t.serialize(true)).toBe(true);
    expect(t.serialize(3.14)).toBe(3.14);
});

test('serialize buffer to base64', ({expect}) => {
    const t = new TestableIPC();
    const buf = Buffer.from([1, 2, 3]);
    expect(t.serialize(buf)).toBe('AQID');
});

test('deserialize primitives', ({expect}) => {
    const t = new TestableIPC();
    expect(t.deserialize(42)).toBe(42);
    expect(t.deserialize('hello')).toBe('hello');
    expect(t.deserialize(true)).toBe(true);
});

test('deserialize blob to buffer', ({expect}) => {
    const t = new TestableIPC();
    const result = t.deserialize({t: 'blob', d: 'AQID'});
    expect(Buffer.isBuffer(result)).toBe(true);
    expect(result).toEqual(Buffer.from([1, 2, 3]));
});

test('serialize then deserialize blob round-trip', ({expect}) => {
    const t = new TestableIPC();
    const original = Buffer.from([0xDE, 0xAD, 0xBE, 0xEF]);
    // Go serializes as {t: 'blob', d: base64}, TS serializes as just base64 string
    // The TS serialize returns base64 string directly for Buffer
    const serialized = t.serialize(original);
    expect(typeof serialized).toBe('string');
});

test('deserialize unknown object throws', ({expect}) => {
    const t = new TestableIPC();
    expect(() => t.deserialize({foo: 'bar'})).toThrow('cannot deserialize');
});

test('serialize unsupported object throws', ({expect}) => {
    const t = new TestableIPC();
    expect(() => t.serialize({foo: 'bar'})).toThrow('cannot serialize');
});

File: lib/golang/serialize_test.go
package golang

import (
	"reflect"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestSerialize(t *testing.T) {
	ipc := &ipcCommon{}

	t.Run("primitives pass through", func(t *testing.T) {
		assert.Equal(t, 42, ipc.serialize(42))
		assert.Equal(t, "hello", ipc.serialize("hello"))
		assert.Equal(t, true, ipc.serialize(true))
		assert.Equal(t, 3.14, ipc.serialize(3.14))
	})

	t.Run("byte slice serializes to blob", func(t *testing.T) {
		data := []byte{0x01, 0x02, 0x03}
		result := ipc.serialize(data)
		m, ok := result.(map[string]any)
		assert.True(t, ok)
		assert.Equal(t, "blob", m["t"])
		assert.Equal(t, "AQID", m["d"]) // base64 of {1,2,3}
	})

	t.Run("empty byte slice serializes to blob", func(t *testing.T) {
		data := []byte{}
		result := ipc.serialize(data)
		m, ok := result.(map[string]any)
		assert.True(t, ok)
		assert.Equal(t, "blob", m["t"])
		assert.Equal(t, "", m["d"])
	})
}

func TestConvType(t *testing.T) {
	ipc := &ipcCommon{}

	t.Run("float64 to int", func(t *testing.T) {
		result := ipc.ConvType(reflect.TypeOf(0), reflect.TypeOf(0.0), float64(42))
		assert.Equal(t, 42, result)
	})

	t.Run("float64 with fractional part stays float", func(t *testing.T) {
		result := ipc.ConvType(reflect.TypeOf(0), reflect.TypeOf(0.0), float64(42.5))
		assert.Equal(t, float64(42.5), result)
	})

	t.Run("string passes through", func(t *testing.T) {
		result := ipc.ConvType(reflect.TypeOf(""), reflect.TypeOf(""), "hello")
		assert.Equal(t, "hello", result)
	})

	t.Run("bool passes through", func(t *testing.T) {
		result := ipc.ConvType(reflect.TypeOf(true), reflect.TypeOf(true), true)
		assert.Equal(t, true, result)
	})
}
